File: flutter_sdk/lib/data/services/client/yalo_chat_client.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'dart:convert';

import 'package:chat_flutter_sdk/src/common/result.dart';
import 'package:chat_flutter_sdk/src/domain/models/yalo_message/yalo_fetch_messages_response.dart';
import 'package:chat_flutter_sdk/src/domain/models/yalo_message/yalo_text_message_request.dart';
import 'package:http/http.dart';
import 'package:logging/logging.dart';

class Action {
  final String name;
  final void Function() action;

  Action({required this.name, required this.action});
}

class YaloChatClient {
  final String name;
  final String flowKey;
  final String userToken;
  final String authToken;
  final String chatBaseUrl;
  final List<Action> actions;
  final Logger log = Logger('YaloChatClient');
  final Client httpClient;

  YaloChatClient({
    required this.name,
    required this.flowKey,
    // FIXME: Remove this one
    required this.authToken,
    required this.userToken,
    Client? httpClient,
  }) : chatBaseUrl = const String.fromEnvironment('YALO_SDK_CHAT_URL'),
       actions = [], httpClient = httpClient ?? Client();

  void registerAction(String actionName, void Function() action) {
    actions.add(Action(name: actionName, action: action));
  }

  // Sends a yalo text message to the upstream chat service
  Future<Result<Unit>> sendTextMessage(YaloTextMessageRequest request) async {
    try {

      final response = await httpClient.post(
        Uri.parse('$chatBaseUrl/inbound_messages'),
        headers: {
          'content-type': 'application/json',
          'x-user-id': userToken,
          'x-channel-id': flowKey,
          'authorization': 'Bearer $authToken',
        },
        body: jsonEncode(request.toJson()),
      );

      if (response.statusCode == 200) {
        return Result.ok(Unit());
      } else {
        return Result.error(
          Exception('Failed to send message: ${response.statusCode}'),
        );
      }
    } on Exception catch (e) {
      return Result.error(e);
    }
  }

  // Fetches messages from the chat service, the "since" argument
  // is a unix timestamp.
  Future<Result<List<YaloFetchMessagesResponse>>> fetchMessages(
    int since,
  ) async {
    try {
      final baseUrl = '$chatBaseUrl/messages';
      final queryParams = {'since': '$since'};
      final uri = Uri.parse(baseUrl).replace(queryParameters: queryParams);
      final response = await httpClient.get(
        uri,
        headers: {
          'x-user-id': userToken,
          'x-channel-id': flowKey,
          'authorization': 'Bearer $authToken',
        },
      );
      if (response.statusCode == 200) {
        final data = YaloFetchMessagesResponse.fromJsonList(
          jsonDecode(response.body),
        );
        return Result.ok(data);
      }
      return Result.error(Exception('Error fetching messages $response'));
    } on Exception catch (e) {
      return Result.error(e);
    }
  }
}

File: typescript-sdk/packages/core/src/data/client/yalo-chat-client.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import { err, ok, type Result } from '../../common/result.js';
import type {
  YaloFetchMessagesResponse,
  YaloTextMessageRequest,
} from '../../domain/yalo-message.js';

export interface Action {
  name: string;
  action: () => void;
}

export interface YaloChatClientConfig {
  name: string;
  flowKey: string;
  userToken: string;
  authToken: string;
  chatBaseUrl: string;
}

export class YaloChatClient {
  readonly name: string;
  readonly flowKey: string;
  readonly userToken: string;
  readonly authToken: string;
  readonly chatBaseUrl: string;
  readonly actions: Action[] = [];

  constructor(config: YaloChatClientConfig) {
    this.name = config.name;
    this.flowKey = config.flowKey;
    this.userToken = config.userToken;
    this.authToken = config.authToken;
    this.chatBaseUrl = config.chatBaseUrl;
  }

  registerAction(name: string, action: () => void): void {
    this.actions.push({ name, action });
  }

  /** Sends a text message to the upstream chat service. */
  async sendTextMessage(request: YaloTextMessageRequest): Promise<Result<void>> {
    try {
      const response = await fetch(`${this.chatBaseUrl}/inbound_messages`, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          'x-user-id': this.userToken,
          'x-channel-id': this.flowKey,
          authorization: `Bearer ${this.authToken}`,
        },
        body: JSON.stringify(request),
      });

      if (response.ok) {
        return ok(undefined);
      }
      return err(new Error(`Failed to send message: ${response.status}`));
    } catch (e) {
      return err(e instanceof Error ? e : new Error(String(e)));
    }
  }

  /**
   * Fetches messages from the chat service.
   * @param since Unix timestamp in seconds — returns messages since this time
   */
  async fetchMessages(since: number): Promise<Result<YaloFetchMessagesResponse[]>> {
    try {
      const url = new URL(`${this.chatBaseUrl}/messages`);
      url.searchParams.set('since', String(since));

      const response = await fetch(url.toString(), {
        headers: {
          'x-user-id': this.userToken,
          'x-channel-id': this.flowKey,
          authorization: `Bearer ${this.authToken}`,
        },
      });

      if (response.ok) {
        const data = (await response.json()) as YaloFetchMessagesResponse[];
        return ok(data);
      }
      return err(new Error(`Error fetching messages: ${response.status}`));
    } catch (e) {
      return err(e instanceof Error ? e : new Error(String(e)));
    }
  }
}
